ocaml/xapi-aux/throttle.mli

```diff
@@ -27,28 +27,33 @@ module Batching : sig
   (** batching delay configuration *)
   type t
 
+  (** internal argument for recursive calls *)
+  type arg = private float
+
   val make : delay_before:float -> delay_between:float -> t
   (** [make ~delay_before ~delay_between] creates a configuration,
     where we delay the API call by [delay_before] once,
     and then with [delay_between] between each recursive call.
    *)
 
-  val with_recursive : t -> (('a -> 'b) -> 'a -> 'b) -> 'a -> 'b
+  val with_recursive : t -> ((arg -> 'b) -> arg -> 'b) -> unit -> 'b
```

Comment:
I think the float argument should be hidden from the users of the call, this can be done using partial application, what this avenue avoided for some particular reason?

```ocaml
  val with_recursive : t -> ((unit -> 'a) -> unit -> 'a) -> unit -> 'a

  let with_recursive config f () =
    let rec self arg =
      let arg = Float.min config.delay_between (arg *. 2.) in
      perform_delay ~yield:false arg ;
      (f [@tailcall]) (self arg)
    in
    let self0 arg = f (self arg) in
    perform_delay ~yield:true config.delay_before ;
    f (self0 (config.delay_between /. 16.)) ()

```

Reply:
I think might have to delay the evaluation there, otherwise this crashes due to infinite recursion.

Reply:
```

Reply:
This might work:
```ocaml
  let with_recursive config f () =
    let rec self arg () =
      let arg = Float.min config.delay_between (arg *. 2.) in
      perform_delay ~yield:false arg ;
      (f [@tailcall]) (self arg)
    in
    let self0 () = (f [@tailcall]) (self (config.delay_between /. 16.)) in
    perform_delay ~yield:true config.delay_before ;
    f self0
```

Reply:
Replaced with a slight variation that also preserves the `'a` argument that `f` could use for its own purposes:
```ocaml
  let with_recursive_loop config f =
    let rec self arg input =
      let arg = span_min config.delay_between Mtime.Span.(2 * arg) in
      perform_delay arg ;
      (f [@tailcall]) (self arg) input
    in
    let self0 arg input = (f [@tailcall]) (self arg) input in
    perform_delay config.delay_before ;
    f (self0 config.delay_initial)
```